File: src/domain/workspace.js
import { Result, id, maxLength, now, required } from '../shared/result.js';
const defaults = { name: 'Meu Financeiro', currency: 'BRL', language: 'pt-BR', timezone: 'America/Sao_Paulo' };
const slugify = (s) => s.normalize('NFD').replace(/[\u0300-\u036f]/g,'').toLowerCase().replace(/[^a-z0-9]+/g,'-').replace(/(^-|-$)/g,'');
export class Workspace {
  static createForUser(userId, input = {}) {
    const data = { ...defaults, ...input };
    const errors = [required(userId,'userId'), required(data.name,'name'), maxLength(data.name,120,'name'), required(data.currency,'currency'), required(data.language,'language'), required(data.timezone,'timezone')].filter(Boolean);
    if (errors.length) return Result.fail(errors);
    const at = now();
    return Result.ok({ id: id('wks'), userId, name: data.name, slug: slugify(`${data.name}-${userId}`), currency: data.currency, language: data.language, timezone: data.timezone, createdAt: at, updatedAt: at, deletedAt: null, events: ['WorkspaceCreated'] });
  }
  static archive(workspace) { return Result.ok({ ...workspace, deletedAt: now(), events: [...(workspace.events ?? []),'WorkspaceArchived'] }); }
}

File: backend/FloraFinance.Domain/Workspaces/Workspace.cs
using FloraFinance.Domain.Common;

namespace FloraFinance.Domain.Workspaces;

public sealed class Workspace : AggregateRoot
{
    private Workspace() { }
    private Workspace(Guid userId, WorkspaceName name, WorkspaceSlug slug, Currency currency, string language, string timezone)
    {
        Id = Guid.NewGuid();
        UserId = userId;
        Name = name;
        Slug = slug;
        Currency = currency;
        Language = language;
        Timezone = timezone;
        CreatedAt = DateTimeOffset.UtcNow;
        Raise(new WorkspaceCreated(Id, UserId));
    }

    public Guid UserId { get; private set; }
    public WorkspaceName Name { get; private set; } = null!;
    public WorkspaceSlug Slug { get; private set; } = null!;
    public Currency Currency { get; private set; } = null!;
    public string Language { get; private set; } = null!;
    public string Timezone { get; private set; } = null!;
    public bool IsArchived => DeletedAt.HasValue;

    public static Result<Workspace> CreateInitial(Guid userId)
    {
        if (userId == Guid.Empty) return Result<Workspace>.Failure(new Error("Workspace.UserRequired", "Usuário obrigatório."));
        var name = WorkspaceName.Create("Meu Financeiro");
        var currency = Currency.Create("BRL");
        if (name.IsFailure || currency.IsFailure) return Result<Workspace>.Failure([.. name.Errors, .. currency.Errors]);
        return Result<Workspace>.Success(new Workspace(userId, name.Value!, WorkspaceSlug.From(name.Value!.Value, userId), currency.Value!, "pt-BR", "America/Sao_Paulo"));
    }

    public Result Rename(string name)
    {
        var parsed = WorkspaceName.Create(name);
        if (parsed.IsFailure) return Result.Failure([.. parsed.Errors]);
        Name = parsed.Value!;
        UpdatedAt = DateTimeOffset.UtcNow;
        Raise(new WorkspaceUpdated(Id));
        return Result.Success();
    }

    public void Archive()
    {
        DeletedAt = DateTimeOffset.UtcNow;
        Raise(new WorkspaceArchived(Id));
    }
}

public sealed record WorkspaceName
{
    private WorkspaceName(string value) => Value = value;
    public string Value { get; }
    public static Result<WorkspaceName> Create(string? value) => string.IsNullOrWhiteSpace(value) || value.Length > 120
        ? Result<WorkspaceName>.Failure(new Error("Workspace.NameInvalid", "Nome obrigatório com no máximo 120 caracteres."))
        : Result<WorkspaceName>.Success(new WorkspaceName(value.Trim()));
}

public sealed record WorkspaceSlug(string Value)
{
    public static WorkspaceSlug From(string name, Guid userId) => new($"{Slugify(name)}-{userId:N}"[..Math.Min(120, $"{Slugify(name)}-{userId:N}".Length)]);
    private static string Slugify(string value) => string.Join('-', value.Normalize().ToLowerInvariant().Split(' ', StringSplitOptions.RemoveEmptyEntries));
}

public sealed record WorkspaceCreated(Guid WorkspaceId, Guid UserId) : IDomainEvent { public Guid EventId { get; } = Guid.NewGuid(); public DateTimeOffset OccurredAt { get; } = DateTimeOffset.UtcNow; }
public sealed record WorkspaceUpdated(Guid WorkspaceId) : IDomainEvent { public Guid EventId { get; } = Guid.NewGuid(); public DateTimeOffset OccurredAt { get; } = DateTimeOffset.UtcNow; }
public sealed record WorkspaceArchived(Guid WorkspaceId) : IDomainEvent { public Guid EventId { get; } = Guid.NewGuid(); public DateTimeOffset OccurredAt { get; } = DateTimeOffset.UtcNow; }
